server/auto_updater/embedding_bin_updater.py

```diff
@@ -0,0 +1,307 @@
+        paper_embeddings_file - the file that contains papers and their respective embeddings
+        pca_axes_file- the file containing the PC axes
+        paper_landscape_json_file - the file containing the json file the front end uses
+    """
+
+    # Grab the PCA axes and Word counts
+    # PCA
+    pca_axes_df = pd.read_csv(pca_axes_file, sep="\t")
+
+    # Read embeddings into dictionary
+    bin_centroid = defaultdict(dict)
+    if Path(paper_embeddings_file).suffix == ".xz":
+        infile = lzma.open(paper_embeddings_file, "rt")
+    else:
+        infile = open(paper_embeddings_file, "r")
+
+    reader = csv.DictReader(infile, delimiter="\t")
+
+    for line_num, line in tqdm.tqdm(enumerate(reader)):
+        if line_num == 0:
+            dim = len(line) - 2
+
+        bin_key = doc_bin_mapper[line["document"]]
+        if bin_key not in bin_centroid:
+            bin_centroid[bin_key] = {
+                "vector": np.array([float(line[f"feat_{idx}"]) for idx in range(dim)]),
+                "counter": 1,
+                "journal": Counter([line["journal"]]),
+            }
+
+        else:
+            bin_centroid[bin_key]["counter"] += 1
+            bin_centroid[bin_key]["vector"] += np.array(
+                [float(line[f"feat_{idx}"]) for idx in range(dim)]
+            )
+            bin_centroid[bin_key]["journal"].update(Counter([line["journal"]]))
+
+    infile.close()
+
+    # Word Counts
+    background_sum = np.sum(list(background_dict.values()))
+
+    # data records
+    bin_stat_records = []
+
+    # Iterate through each bin
+    for squarebin_id in tqdm.tqdm(word_bin_dict):
+
+        # Calculate bin token odds
+        bin_sum = np.sum(list(word_bin_dict[squarebin_id].values()))
+        word_odds_ratios = get_odds_ratio(
+            word_bin_dict[squarebin_id], background_dict, background_sum, bin_sum
+        )
+        word_odds_ratios = sorted(
+            word_odds_ratios, key=lambda x: x["odds_ratio"], reverse=True
+        )
+
+        # Calculate pca enrichment
+        bin_vector = (
+            bin_centroid[squarebin_id]["vector"] / bin_centroid[squarebin_id]["counter"]
+        )
+        pca_sim_df = get_pca_sim([bin_vector], pca_axes_df.values)
+
+        # Write stats as records
+        bin_stat_records.append(
+            {
+                "bin_id": squarebin_id,
+                "pc": pca_sim_df.to_dict(orient="records"),
+                "count": bin_centroid[squarebin_id]["counter"],
+                "journal": dict(bin_centroid[squarebin_id]["journal"].items()),
+                "bin_odds": word_odds_ratios[:20],
+            }
+        )
+
+    # Write updated file to disk
+    square_plot_df = pd.read_json(paper_landscape_json_file)
+    (
+        square_plot_df.merge(pd.DataFrame.from_records(bin_stat_records), on="bin_id")
+        .reset_index(drop=True)
+        .to_json(paper_landscape_json_file, orient="records", lines=False)
```

Comment:
Just want to confirm: so `paper_landscape_json_file` is created by `generate_SAUCIE_coordinates()` earlier, then modified here?

Reply:
Correct